packages/tools/devtools/devtools-core/src/data-visualization/SharedTreeHelpers.ts

```diff
@@ -0,0 +1,204 @@
+import { type VisualChildNode, VisualNodeKind, type VisualValueNode } from "./VisualTree.js";
+
+/**
+ * TODO
+ */
+export function visualRepresentationMapper(tree: VisualSharedTreeNode): VisualChildNode {
+	if ("value" in tree) {
+		const result: VisualValueNode = {
+			value: tree.value,
+			nodeKind: VisualNodeKind.ValueNode,
+			metadata: { allowedTypes: tree.schema.allowedTypes },
+		};
+		return result;
+	} else {
+		// Handling SharedTreeNode
+		const children: Record<string, VisualChildNode> = {};
+
+		for (const [key, value] of Object.entries(tree.fields)) {
+			const child = visualRepresentationMapper(value);
+			children[key] = child;
+		}
+
+		return {
+			children,
+			nodeKind: VisualNodeKind.TreeNode,
+			metadata: { allowedTypes: tree.schema.allowedTypes },
+		};
+	}
+}
+
+/**
+ * Helper function to generate the allowed fields & types for the fields of the tree.
+ */
+function allowedTypesHelper(schema: ObjectNodeStoredSchema): string {
+	let result = "";
+
+	for (const [fieldKey, treeFieldStoredSchema] of schema.objectNodeFields) {
+		/**
+		 * Set of allowed tree types {@link TreeTypeSet}.
+		 */
+		const fieldTypes = treeFieldStoredSchema.types;
+
+		let fieldAllowedType = `${fieldKey} : `;
+
+		/**
+		 * If not specified, types are unconstrained.
+		 */
+		if (fieldTypes === undefined) {
+			fieldAllowedType += "any";
+		} else {
+			for (const type of fieldTypes) {
+				fieldAllowedType += `${type} | `;
+			}
+		}
+
+		/**
+		 * Slice the trailing ` | ` from the `fieldAllowedType`.
+		 */
+		fieldAllowedType = `${fieldAllowedType.slice(0, -3)}, `;
+
+		result += fieldAllowedType;
+	}
+
+	/**
+	 * Slice the trailing `, ` from the `result`.
+	 */
+	result = result.slice(0, -2);
+
+	return `{ ${result} }`;
+}
+
+function allowedTypesMapHelper(
+	fields: FieldMapObject<JsonableTree> | undefined,
+	schema: MapNodeStoredSchema,
+): string {
+	// TODO: Write this function.
+	const result = "";
+	return result;
+}
+
+function leafNodeStoredSchemaHelper(
+	tree: JsonableTree,
+	schema: LeafNodeStoredSchema,
+): SharedTreeLeafNode {
+	return {
+		schema: {
+			allowedTypes: JSON.stringify(schema.leafValue),
+		},
+		value: JSON.stringify(tree.value), // TODO: this needs to be `await visualizeChildData(tree.value), otherwise we won't handle Fluid Handles correctly
+	};
+}
+
+function objectNodeStoredSchemaHelper(
+	tree: JsonableTree,
+	schema: ObjectNodeStoredSchema,
+	contentSnapshot: SharedTreeContentSnapshot,
+): VisualSharedTreeNode {
+	const treeFields = tree.fields;
+
+	if (treeFields === undefined || Object.keys(treeFields).length === 0) {
+		// TODO: what does this case mean?
+		return {
+			schema: { name: tree.type, allowedTypes: allowedTypesHelper(schema) },
+			fields: {},
+		};
+	}
+
+	const fields: Record<string | number, VisualSharedTreeNode> = {};
+
+	if (
+		Object.keys(treeFields).length === 1 &&
+		Object.prototype.hasOwnProperty.call(treeFields, "")
+	) {
+		const children = treeFields[""]; // TODO: Fail otherwise.
```

Comment:
And let's add a comment explaining that the empty key indicates an array field, and that we hide away the level of indirection in the visual output because it's not useful to developers using the devtools.